R/utils.R

```diff
@@ -103,46 +103,72 @@ Warn = function(msg, ...) rlang::warn(break_str(msg, init = "Warning: "), ...)
+           class = "epiprocess__assert_sufficient_f_args__mandatory_f_args_passed_to_f_dots",
+           epiprocess__f = f,
+           epiprocess__mandatory_f_args_in_f_dots = mandatory_f_args_in_f_dots)
     }
   } else {
-    if (length(arg_names) < n_mandatory_f_args) {
-      Abort(sprintf("`f` must take at least %s arguments", n_mandatory_f_args),
-          class="epiprocess__assert_sufficient_f_args__f_needs_min_args",
-          epiprocess__f = f,
-          epiprocess__arg_names = arg_names)
+    if (length(args_names) < n_mandatory_f_args + rlang::dots_n(...)) {
+      if (rlang::dots_n(...) == 0L) {
+        # common case; try for friendlier error message
+        Abort(sprintf("`f` must take at least %s arguments", n_mandatory_f_args),
+              class = "epiprocess__assert_sufficient_f_args__f_needs_min_args",
+              epiprocess__f = f)
+      } else {
+        # less common; highlight that they are (accidentally?) using dots forwarding
+        Abort(sprintf("`f` must take at least %s arguments plus the %s arguments forwarded through `epi[x]_slide`'s `...`", n_mandatory_f_args, rlang::dots_n(...)),
+              class = "epiprocess__assert_sufficient_f_args__f_needs_min_args_plus_forwarded",
+              epiprocess__f = f)
+      }
     }
   }
-  # If `f` has fewer than n_mandatory_f_args before `...`, then we only need
-  # to check those args for defaults.
-  n_args_before_dots = min(
-    ifelse(exists("arg_names_before_dots"), length(arg_names_before_dots), n_mandatory_f_args),
-    n_mandatory_f_args
-  )
-  if ( any(map_lgl(args_f[seq(n_args_before_dots)], ~!is_missing(.x))) ) {
-      Abort(sprintf("Some of `f`'s first %s arguments use defaults that would be overridden when calling `f` to run the computation", n_args_before_dots),
-          class="epiprocess__assert_sufficient_f_args__required_args_contain_defaults",
-          epiprocess__f = f,
-          epiprocess__args_f = args_f)
+  # Check for args with defaults that are filled with mandatory positional
+  # calling args. If `f` has fewer than n_mandatory_f_args before `...`, then we
+  # only need to check those args for defaults. Note that `dots_i - 1L` is
+  # length 0 if `f` doesn't accept `...`.
+  n_remaining_args_for_default_check = min(c(dots_i - 1L, n_mandatory_f_args))
+  default_check_args = remaining_args[seq_len(n_remaining_args_for_default_check)]
+  default_check_args_names = names(default_check_args)
+  has_default_replaced_by_mandatory = map_lgl(default_check_args, ~!is_missing(.x))
+  if (any(has_default_replaced_by_mandatory)) {
+    mandatory_args_replacing_defaults =
+      mandatory_f_args_labels[seq_len(sum(has_default_replaced_by_mandatory))]
```

Comment:
issue: `mandatory_args_replacing_defaults` should be defined positionally. The `seq_len(sum())` is causing both cases `f(x=1, g)` and `f(x, g=1)` to warn that "`epi[x]_slide` would pass the window data" to the arg with the default. But the second case would pass the group key arg instead.

```suggestion
      mandatory_f_args_labels[has_default_replaced_by_mandatory]
```

Reply:
This actually had a slight bug: `mandatory_f_args_labels` isn't necessarily of the same length as `has_default_replaced_by_mandatory`.  Should be fixed now.  Would have been easier to spot if we had better names for some of these variables, but I can't think of better names at the moment.